### DISCUSSION TITLE
[NRG Request] Private Shared States with up to US$100k in Grants

### DISCUSSION THREAD

**Discussion by Savio-Sou** [noir-lang/noir#6289] [NRG#2 Private Shared States]
Zero-Knowledge Proofs (ZKP) unlock the ability for users to compute over private data, but only on data they individually own. 

Private shared states (PSS) can enable valuable use cases, such as brokers pairing dark pool orders, dating apps matching mutual interests, and game servers computing fog of wars, while keeping data obscure to all users until specific conditions are met.

[Aztec Labs](https://aztec-labs.com/), is a strong believer in the value of PSS, and is providing up to US$100,000 in grants to explore innovations with PSS, powered by Noir.

Advancements in mix-and-matching [programmable cryptography](https://0xparc.org/blog/programmable-cryptography-1) technologies are enabling many possible ways to achieve cryptographic PSS (versus trusted middleman pseudo-PSS), including:

- Noir <> Fully Homomorphic Encryption (FHE)
- Noir <> Multi-Party Computation (MPC)
- Noir <> Trusted Execution Environments (TEE)

This request is unbiased on the technologies paired with Noir to achieve PSS. Proposals building applications (including MVPs), developer tooling, technical or cryptography research, are all welcome as long as their relevance to PSS is clear.

Proposals should take into account that relevant primitives could be missing or under-optimized in Noir (e.g. libraries, tools). Implementing, optimizing and open-sourcing these should be considered in-scope.

## Submission

To ensure consistency and facilitate the review process, please follow the following submission format:

- **Title:** A concise, descriptive title for your proposal
- **Summary:** A high-level summary of your proposal (≤200 words)
- **Methodology:** Explain the research approach and design details
- **Assumptions:** Walk through assumptions (e.g. trust, privacy, security) of the design
- **Timeline and Deliverables:** Describe project milestones and what is expected to be achieved when
- **Team:** List team member(s) that will work on the initiative, relevant experience and contacts
- **Start Date:** Specify preferred start date (e.g. 11 November 2024)
- **Questions:** List outstanding questions, if any

Submit a proposal by creating a new post in [this category](https://github.com/orgs/noir-lang/discussions/categories/-nrg-2-private-shared-states) using the format by October 25th, 2024.

## Requirements

Implementations of proposals should observe the following requirements. If your proposal could not meet all requirements, please highlight the missing requirements accordingly.

- Final implementation MUST contain components written in Noir
- Final implementation MUST be well documented
- Final implementation MUST be open sourced under a permissive license (e.g. MIT)
- Final implementation SHOULD be thoroughly tested
- Project proposals SHOULD size for approximately 3 months of works
- Final implementation CAN contain components written with Aztec.nr

## Evaluation

Proposals will be evaluated on five categories based on the following criteria:

- Adherence to requirements
- Innovativeness
- Quality
- Feasibility
- Vision to build with Noir long term

With Aztec Labs’ sponsorship, proposals that demonstrate a clear vision for end-user adoption and long-term development on Aztec will receive preferential consideration.

Design assumptions such as trust, privacy, and security of proposed projects will also be taken into account for this NRG specifically, .

Selected proposals will be announced on GitHub Discussions by November 8th, 2024.

If you would like to sponsor additional prizes for this research campaign, please [message Savio on Signal](https://signal.me/#eu/bx9Z-C5Ax8Kbf9WuxuLDDaDqYOqRz4m_fY1_XKq2hiC5vcdEMitN6Zhfdz-qRC8K).

## Resources

- Conceptual understanding
  - Independent <> Mediated <> Global Computation https://knot-group.github.io/wiki/building/framework/index.html
  - Navigating Privacy on Public Blockchains: https://wdai.us/posts/navigating-privacy/
- Noir
  - Docs https://noir-lang.org/docs/
  - Resource registry https://github.com/noir-lang/awesome-noir/
  - Discord https://discord.gg/JtqzkdeQ6G for language specific problems and discussions
- ZKP <> MPC
  - [Paper] Collaborative zk-SNARKs: https://eprint.iacr.org/2021/1530
  - [Blog] Taceo - Private Proof Markets using MPC: https://blog.taceo.io/private-proof-markets-using-mpc/
  - [Blog] Taceo - coSNARK in Noir explainer: https://blog.taceo.io/co-noir-intro/
  - [Tooling] Taceo - coSNARK in Noir repository: https://github.com/TaceoLabs/collaborative-circom/tree/main/co-noir
- ZKP <> TEE
  - [Docs] Marlin - Kalyso with Noir tutorial (TEE proofs with private inputs): https://docs.marlin.org/user-guides/kalypso/tutorials/noir-circuits
  - [Tooling] Marlin - Noir proving enclave: https://github.com/marlinprotocol/noir_prover_enclave/
- ZKP <> FHE
  - [Blog] Christian K - Verifiable Fully Homomorphic Encryption: https://cknabs.github.io/post/vfhe/
  - [Tooling] Christian K - zkFHE in different stacks: https://github.com/zkFHE

## FAQ

**Q1. How will proposals be chosen?**

The selection committee will consist of representatives from [Aztec Labs](https://aztec-labs.com/) and possibly other teams, who will evaluate and select proposals following the criteria and deadlines above.

**Q2. Who can submit proposals?**

Researchers, developers, and technology enthusiasts with an interest in Noir research are encouraged to submit their proposals.

**Q3. Can I submit more than one proposal?**

Yes, you can submit multiple proposals if you have different ideas.

**Q4. Why should I submit a proposal early?**

That way sponsors and the community could share with you any early feedback applicable.

The earlier the submission, the higher the chance sponsors and the community could review your proposal and share any feedback applicable before the submission deadline.

**Q5. Can I update my proposal after submission?**

Yes, edits are welcome up to the submission deadline.

**Q6. What if my proposal does not fully meet the requirements?**

You are encouraged and welcome to submit your proposal and participate in the discussion still. Your ideas could contribute valuable insights and help shape further research.

**Q7. What should I do if I have a question not covered above?**

Please comment your questions below in this NRG Request.

**Q8. How will grants be distributed?** 

Up to a total of two proposals will be selected and each selected proposal would be supported with US$50,000: 

Distribution timeline:

- $10,000 on publishing 1st month’s progress report
- $20,000 on publishing 2nd month’s progress report
- $20,000 on publishing 3rd month’s final report

## Acknowledgements

Thank you to [Aztec Labs](https://aztec-labs.com/) for sponsoring the grants for selected proposals.

Thank you to the [Marlin](https://www.marlin.org/) and [Taceo](https://taceo.io/) teams for supporting NRG#2 applicants with infrastructural tooling and technical advisory.

Thank you to [Arnaud Schenk](https://x.com/_ArnaudS_), [Rahul Kothari](https://x.com/omw_to_the_moon), and [Wei Dai](https://x.com/_weidai) for reviewing the draft of this request.

## DISCLAIMER

_The information set out herein is for discussion purposes only and does not represent any binding indication or commitment by Aztec Labs and its employees to take any action whatsoever, including relating to the structure, any potential operation and/or features of the Aztec protocol, the protocol roadmap or the Noir language. In particular: (i) nothing in these posts is intended to create any contractual or other form of legal relationship with Aztec Labs or third parties who engage with such posts (including, without limitation, by submitting a proposal or responding to posts), (ii) by engaging with any post, the relevant persons are consenting to Aztec Labs’ use and publication of such engagement and related information on an open-source basis (and agree that Aztec Labs will not treat such engagement and related information as confidential), and (iii) Aztec Labs is not under any duty to consider any or all engagements, and that consideration of such engagements and any decision to award grants or other rewards for any such engagement is entirely at Aztec Labs’ sole discretion. Please do not rely on any information on this forum for any purpose - the development, release, and timing of any products, features or functionality remains subject to change and is currently entirely hypothetical. Nothing on this forum should be treated as an offer to sell any security or any other asset by Aztec Labs or its affiliates, and you should not rely on any forum posts or content for advice of any kind, including legal, investment, financial, tax or other professional advice._

**Comment by aisconnolly** [comment]
This is such a nice initiative Savio, and well deserved for Noir. 

If any one is interested in building in the **ZKP <> MPC** direction, the TACEO team will hold office hours in [our discord](https://discord.gg/BGGhtwuQ) on 
* Thursday 17th October at 17:00 CET (11:00 EST)
* Thursday 24th October at 17:00 CET (11:00 EST)

Stop by if you have questions or to talk through our coSNARK tooling for Noir!

**Comment by gamgote** [comment]
This is a very informative post! 

If developers would like to explore the use of TEEs to compute over shared state, we'll be hosting an office hour at the time below. This is a great opportunity to learn more about TEEs, [Oyster](marlin.org/oyster) and [Kalypso](marlin.org/kalypso) in general. 

When: 24th October at 12PM GMT
Where: Join us via Google Meet: https://meet.google.com/bsr-gfqw-snv

Also feel free to DM us on Twitter (@MarlinProtocol) any time. Happy to provide credits and support to deploy code on [Oyster](marlin.org/oyster) TEEs and create markets on [Kalypso](marlin.org/kalypso).

**Comment by Savio-Sou** [comment]
On popular request, we are updating the proposal deadlines as follows:
- Deadline to create proposals: October 25th, 2024
- Deadline to update proposals: October 28th, 2024

Good luck teams!
